---
title: "benchpro strain"
author: Joachim Fritscher
date: September 07, 2023
output: 
  html_document:
  toc: true
  toc_float:
  toc_collapsed: true
  toc_depth: 3
  number_sections: true
  theme: lumen
params:
  meta: meta
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width=12)
library(vegan)
library(ape)
library(ggtree)
library(ggtreeExtra)
library(phytools)
library(dplyr)
library(ggplot2)
library(tidyr)
library(forcats)
library(knitr)
library(plotly)
library(reshape2)
library(phangorn)
library(cowplot)
library(ggpubr)
library(ggExtra)
# library(tidyverse)
```


eval_global <- FALSE

```{r message=FALSE, warning=FALSE, include=FALSE}
source("strain_functions.R")
source("strain_plots.R")
```


```{r, echo=FALSE, warnings=FALSE, message=FALSE}
META_GROUP_COLUMN <- "genome"
GOLD_STD_TOOL <- "gold_std"

map.file <- "/usr/users/QIB_fr017/fritsche/Projects/benchpro/data/SPECIES46/map2.csv"
map.file <- "/usr/users/QIB_fr017/fritsche/Projects/benchpro/data/SPECIES46/MSSS200/map.csv"
map.file <- "/usr/users/QIB_fr017/fritsche/Projects/benchpro/data/SPECIES46/MSSS200C/map.csv"
map.file <- "/usr/users/QIB_fr017/fritsche/Projects/benchpro/data/SPECIES46/MSSS200C/map2.csv"
map.file <- "/usr/users/QIB_fr017/fritsche/Projects/benchpro/data/SPECIES46/MSSS200C/map4.csv"
# map.file <- "/usr/users/QIB_fr017/fritsche/Projects/benchpro/data/SPECIES46/MSSS200C/map4_test.csv"

map <- load_map(map.file) %>% filter(Tool %in% c("StrainPhlan4", "gold_std"))
meta.dict <- get_meta_dict(map)

# If gold standards do not exist
generate_gold_stds(map, meta.dict)

tree.dict <- get_tree_dict(map)
tree_dict_shared_tips <- get_tree_dict_shared_tips(map, tree.dict)
shared_species <- species_for_tools(map)


data <- load_all(map)
```


<details>
  <summary>How monophyly is assessed?</summary>
  ### Monophyly
  A monophyletic group or clade contains an ancestral taxon and all its descendants. When given a strain-level tree for a species, each tip represents a sample. The tree then assesses how similar the strains for that species are between samples. Whenever two samples contain the same strain (see definition of strain below) we expect them to cluster together in the tree with no other samples (with potentially very similar strains) in-between. Monophyly as a metric is especially useful in a time-series data setting as two adjacent samples of an individual are often expected to carry the identical strain of a species. In simulated samples where we know which samples contain reads from the same genome, we can calculate how well strain-level tree reflects these monophyletic relationships. In real time-series data we still expect that higher values of monophyly indicate a better performance.

  ### Definition of strain
  The definition of "strain" we use here is based on genome similarity (average nucleotide identity (ANI) on whole or parts of genomes). Two samples carry the same bacterial strain if they are in direct, recent ancestral relationship. In a host microbiome setting this would happen through a) Time series samples of a single patient where the same bacterial lineage is present in two adjacent samples or b) If there has been a transmission from one patient to another so that they carry the same strain. Two samples are from the same patient with one year difference and the "same" bacterium is still present one year after, then both samples carry the same strain.

  ### Calculating monophyly
  Monophyly here is calculated as the ratio of samples that are supposed to carry the same strain (and are present in the tree) with respect to the number of tips in the smallest possible clade that contains all of these samples.
    <img width="50%" height="50%" style="margin: 10px" src="images/monophyly.png"/>
</details>
<br>

# Analysis {.tabset}

## Monophyly Analysis


```{r, echo=FALSE, warnings=FALSE, message=FALSE, eval=TRUE}

if (length(data$tree.dict.shared.tips) > 0) {
  monophyly.df <- monophylies(data$tree.dict.shared.tips, data$map, data$meta.dict, META_GROUP_COLUMN) %>% 
    left_join(data$map, by=MAP_NAME_COL) %>% 
    rm_duplicate_suffix()
  monophyly.cov.df <- monophylies_per_coverage(data$tree.dict.shared.tips, data$map, data$meta.dict, META_GROUP_COL) %>% 
    left_join(data$map, by=MAP_NAME_COL) %>% 
    rm_duplicate_suffix()
  
  mono.plot <- monophyly.df %>% 
    filter(Tool != GOLD_STD_TOOL) %>%
    plot_monophyly(title="Monophyly compared (only on shared tips)")
  mono.plot.shared <- monophyly.df %>% 
    filter(Tool != GOLD_STD_TOOL & Species %in% data$shared_species) %>%
    plot_monophyly(title="Monophyly compared (only on shared tips and shared species)")
  mono.plot.summary <- monophyly.df %>% 
    filter(Tool != GOLD_STD_TOOL) %>% 
    plot_monophyly_summary(title="Monophyly summary (only on shared tips)")
  mono.plot.summary.shared <- monophyly.df %>% 
    filter(Tool != GOLD_STD_TOOL & Species %in% data$shared_species) %>%
    plot_monophyly_summary(title="Monophyly summary (only on shared tips and shared species)")
  mono.plot.cov.summary.shared <- plot_monophyly_per_coverage(monophyly.cov.df %>% 
    filter(TreeTool == "iqtree" & Tool != GOLD_STD_TOOL & Species %in% data$shared_species), title="Monophyly summary (only on shared tips and shared species)")
  
  print(mono.plot)
  print(mono.plot.shared)
  print(mono.plot.summary)
  print(mono.plot.summary.shared)
  print(mono.plot.cov.summary.shared)
}

# Set up monophyly dataframe
monophyly.df <- monophylies(data$tree.dict, data$map, data$meta.dict, META_GROUP_COL) %>% left_join(data$map, by=MAP_NAME_COL) %>% rm_duplicate_suffix()
monophyly.df$Species <- unlist(lapply(monophyly.df$Species, abbrev_species))
# monophyly.df <- rbind(
#   monophyly.df %>% filter(Tool != GOLD_STD_TOOL),
#   monophyly.df %>% filter(Tool != GOLD_STD_TOOL) %>%
#     mutate(Species = "Summary")
# )

all.meta <- bind_rows(data$meta.dict, .id = "column_label")
all.meta %>% group_by(genome, species) %>%
  summarise(samples = n())

mono.plot.all <- monophyly.df %>%
  filter(Tool != GOLD_STD_TOOL) %>%
  plot_monophyly(title="Monophyly") +
  theme(legend.position="bottom",
        axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
mono.plot.summary <- monophyly.df %>%
  filter(Tool != GOLD_STD_TOOL) %>%
  mutate(Species = "Summary") %>%
  plot_monophyly(title="") +
  theme(legend.position="None",
        axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
mono.plot.combined <- plot_grid(mono.plot.all, mono.plot.summary, rel_widths = c(0.85,0.15))


mono.plot.summary <- monophyly.df %>%
  filter(Tool != GOLD_STD_TOOL) %>% 
  plot_monophyly_summary(title="Monophyly")
mono.plot.sharedsp.summary <- monophyly.df %>%
  filter(Tool != GOLD_STD_TOOL & Species %in% data$shared_species) %>%
  plot_monophyly_summary(title="Monophyly on shared species")


print(mono.plot.all)
print(mono.plot.summary)
print(mono.plot.sharedsp.summary)

n_perfect <- monophyly.df %>% filter(Value==1 & Tool == "StrainPhLAn4") %>% count()
n_imperfect <- monophyly.df %>% filter(Value!=1 & Tool == "StrainPhLAn4") %>% count()
paste("Resolved species at monophyly 1: ", n_perfect/nrow(monophyly.df))
n_perfect_strains <- monophyly.df %>% filter(Value==1 & Tool == "StrainPhLAn4") %>% pull(GroupSize) %>% as.numeric() %>% sum()
n_imperfect_strains <- monophyly.df %>% filter(Value!=1 & Tool == "StrainPhLAn4") %>% pull(GroupSize) %>% as.numeric() %>% sum()
paste("Resolved strains at monophyly 1: ", n_perfect_strains/sum(monophyly.df %>%pull(GroupSize) %>% as.numeric()))

monophyly_summary <- monophyly.df %>% group_by(Tool) %>%
  summarise(total = n(), total_samples=sum(GroupSize), perfect = sum(Value == 1), perfect_samples = sum(GroupSize[Value == 1])) %>%
  mutate(perfect_rate=perfect/total, perfect_sample_rate=perfect_samples/total)

monophyly_species_summary_tool <- monophyly.df %>% 
  filter(Tool != GOLD_STD_TOOL) %>%
  group_by(Tool, Species) %>%
  summarise(n = n(), n_samples=sum(GroupSize), perfect = sum(Value == 1), perfect_samples = sum(GroupSize[Value == 1])) %>%
  mutate(perfect_rate=perfect/n, perfect_sample_rate=perfect_samples/n_samples)
# monophyly_species_summary_gold <- monophyly.df %>% 
#   filter(Tool == GOLD_STD_TOOL) %>%
#   group_by(Species) %>%
#   summarise(n_gold = n(), n_samples_gold=sum(GroupSize))

monophyly_species_summary_gold <- all.meta %>% group_by(species) %>%
  summarise(n_samples_gold=n(), n_gold=length(unique(genome)), mean_coverage=mean(coverage)) %>%
  mutate(Species=gsub(" ", "_", species))

monophyly_strain_summary_gold <- all.meta %>% group_by(species, genome) %>%
  summarise(group_size=n())


monophyly_species_summary_tool <- monophyly_species_summary_tool %>%
  left_join(monophyly_species_summary_gold, by="Species") %>%
  mutate(sensitivity=n/n_gold, sensitivity_sample=n_samples/n_samples_gold)

mean_sd(monophyly_species_summary_tool$sensitivity)

sensitivity.plot <- monophyly_species_summary_tool %>%
  ggplot(aes(x=sensitivity, y=sensitivity_sample, color=Tool, fill=Tool)) +
  geom_point() + 
  theme(legend.position="bottom") +
  xlab("Strain sensitivity") +
  ylab("Sample sensitivity") +
  xlim(c(0,1)) + ylim(c(0,0.5))
ggMarginal(sensitivity.plot, type="boxplot", size=16, groupFill=TRUE)

mean_sd <- function(x, digits=2) {
  paste(round(mean(x), digits), round(sd(x), digits), sep="Â±")
}

group_sizes <- monophyly.df %>% 
  filter(Tool == GOLD_STD_TOOL) %>% pull(GroupSize)
length(group_sizes)
group_sizes %>% mean_sd()



per_species_sample <- monophyly_species_summary_tool %>%
  ggplot(aes(x=sensitivity_sample, y=perfect_sample_rate)) +
  geom_point() +
  geom_smooth(method="lm") +
  stat_cor(method="pearson") +
  ggtitle("Per species sample statistics") +
  xlab("Sample sensitivity") +
  ylab("Percent of samples with perfect Monophyly")
per_species_strain <-monophyly_species_summary_tool %>%
  ggplot(aes(x=sensitivity, y=perfect_rate)) +
  geom_point() +
  geom_smooth(method="lm") +
  stat_cor(method="pearson") +
  ggtitle("Per species strain statistics") +
  xlab("Sample sensitivity") +
  ylab("Percent of strains with perfect Monophyly")
plot <- plot_grid(per_species_sample, per_species_strain)


total_taxa <- sum(as.numeric(monophyly.df$GroupSize))

monophyly.df
```


```{r, echo=FALSE, warnings=FALSE, message=FALSE, eval=FALSE}
print(mono.plot.sharedsp.summary)
```

## Strain Cluster & ANI based Analysis

```{r, echo=FALSE, warnings=FALSE, message=FALSE, eval=TRUE}
if (length(tree_dict_shared_tips) > 0) {
  melts.shared <- pairwise_distances(tree_dict_shared_tips, map, meta.dict, META_GROUP_COL) %>% 
    left_join(map, by=MAP_NAME_COL)
  
  melts.plot <- melts.shared %>%
    filter(Species %in% shared_species) %>% 
    plot_intra_vs_inter(title="Intra genome group vs inter genomegroup distance (shared tips and species)")
  print(melts.plot)
  
  ani.plots.inter.stratified.shared <- melts.shared %>%
    filter(Species %in% shared_species & Tool != GOLD_STD_TOOL) %>% 
    filter(identity == TRUE) %>% 
    ggplot(aes(x=Tool, y=value, fill=Tool)) +
    facet_wrap(~Species) +
    geom_boxplot() + 
    coord_cartesian(ylim=c(0.995, 1)) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  print(ani.plots.inter.stratified.shared)
}

pairwise.all <- pairwise_distances(tree.dict, map, meta.dict, META_GROUP_COL) %>% 
  left_join(map, by=MAP_NAME_COL)

melts.plot <- pairwise.all %>%
  filter(Species %in% shared_species) %>%
  plot_intra_vs_inter(title="Intra genome group vs inter genomegroup distance (shared tips and species)")
print(melts.plot)


ani.plots.inter <- ggplot(pairwise.all %>% filter(identity == TRUE), aes(x=Tool, y=value, fill=TreeTool)) +
  geom_boxplot()
print(ani.plots.inter)

# print(length(tree_dict_shared_tips))
# ani.plots.inter.stratified <- pairwise.all %>% 
#   filter(identity == TRUE & Tool != GOLD_STD_TOOL) %>% 
#   ggplot(aes(x=Tool, y=value, fill=Tool)) +
#   facet_wrap(~Species) +
#   geom_boxplot() + coord_cartesian(ylim=c(0.995, 1)) + 
#   theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
# print(ani.plots.inter.stratified)
```

```{r,echo=FALSE, warnings=FALSE, message=FALSE, eval=TRUE, fig.width=10, fig.height=11}

# print(ani.plots.inter.stratified.shared)
ani.plots.inter.stratified <- pairwise.all %>% 
  filter(identity == TRUE & Tool != GOLD_STD_TOOL) %>% 
  ggplot(aes(x=Tool, y=value, fill=Tool)) +
  facet_wrap(~Species) +
  geom_boxplot() + coord_cartesian(ylim=c(0.995, 1)) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
print(ani.plots.inter.stratified)

duplicate <- function(df, col) {
  return(rbind(df, df %>% mutate(!!col := "all")))
}
# print(ani.plots.inter.stratified.shared)
ani.plots.inter.stratified <- pairwise.all %>% 
  filter(identity & Tool != GOLD_STD_TOOL & sample1 != sample2) %>% 
  group_by(genome1) %>% 
  duplicate("Species") %>%
  slice_min(value, n=1) %>%
  mutate(norm_value=(1-value)/max) %>%
  ggplot(aes(x=Tool, y=norm_value, fill=Tool)) +
  facet_wrap(~Species) +
  geom_boxplot() + 
  ggtitle("Pairwise distances in strain-resolved tree per species") +
  ylab("Normalized distance (lower is better)") +
  labs(subtitle = "Normalized distance: max(pairwise_same_strain) / max(pairwise_all). Pairwise strain means pairwise distance between two samples with the same strain.") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  coord_cartesian(ylim=c(0, 0.25))
print(ani.plots.inter.stratified)
```




```{r, echo=FALSE, warnings=FALSE, message=FALSE, eval=TRUE}
library(tidyquant)

window_size = 50

if (length(data$tree.dict.shared.tips) > 0) {
  monophyly.df <- monophylies(data$tree.dict.shared.tips, data$map, data$meta.dict, META_GROUP_COL) %>% 
    left_join(map, by=MAP_NAME_COL) %>% 
    rm_duplicate_suffix()
  
  target.mono <- monophyly.df %>% filter(TreeTool == "iqtree")
  
  gold_std.pairwise.inter <- pairwise.all %>% 
    complete_pairwise() %>% 
    filter(TreeTool == "iqtree" & !identity & Tool == "gold_std" & sample1 != sample2) %>% 
    group_by(Group) %>% 
    summarise(minsim=min(value), maxsim=max(value), Species=head(Species, 1), Tool=head(Tool, 1)) %>% 
    as.data.frame()
  
  join.gstd <- target.mono %>% 
    left_join(gold_std.pairwise.inter %>% select(-Tool), c("Group", "Species")) %>% 
    filter(Tool != "gold_std") %>% rm_duplicate_suffix()
  
  if (join.gstd %>% filter(Species %in% shared_species) %>% 
      sliding_window_possible(window_size = window_size)) {
    
      # Create plots
    swm1.shared <- join.gstd %>% filter(Species %in% shared_species) %>% 
      sliding_window_monophyly(window_size=window_size) + 
      geom_smooth(method = "loess") +
      ggtitle("Mean monophyly per genome (Shared species, shared tips)")
    
    swm1 <- join.gstd %>% 
      sliding_window_monophyly(window_size) + 
      geom_smooth(method = "loess") +
      ggtitle("Mean monophyly per genome (shared tips)")
    
    # Print plots
    print(swm1.shared)
    print(swm1)
  }
}

# Species monophylies
monophyly.df <- monophylies(tree.dict, map, meta.dict, META_GROUP_COL) %>% 
  left_join(map, by=MAP_NAME_COL) %>% 
  rm_duplicate_suffix()

target.mono <- monophyly.df %>% filter(TreeTool == "iqtree")

gold_std.pairwise.inter <- pairwise.all %>% 
  complete_pairwise() %>% 
  filter(TreeTool == "iqtree" & !identity & Tool == "gold_std" & sample1 != sample2) %>% 
  group_by(Group) %>% 
  summarise(minsim=min(value), maxsim=max(value), Species=head(Species, 1), Tool=head(Tool, 1)) %>% 
  as.data.frame()
# target.melts

join.gstd <- target.mono %>% 
  left_join(gold_std.pairwise.inter %>% select(-Tool), c("Group", "Species")) %>% 
  filter(Tool != "gold_std") %>% rm_duplicate_suffix()

if (join.gstd %>% filter(Species %in% shared_species) %>% 
      sliding_window_possible(window_size = window_size)) {
  swm1.shared <- join.gstd %>% filter(Species %in% shared_species) %>% 
    sliding_window_monophyly(window_size) + 
    geom_smooth(method = "loess") +
      ggtitle("Mean monophyly per genome (Shared species)")
  
  swm1 <- join.gstd %>% 
    sliding_window_monophyly(window_size) + 
    geom_smooth(method = "loess") +
    ggtitle("Mean monophyly per genome")
  
  
  print(swm1.shared)
  print(swm1)
}
```


```{r, echo=FALSE, eval=FALSE, warnings=FALSE, message=FALSE}
# sliding_window_mono <- function(df, window_size) {
#   join.gstd.rank <- df %>% mutate(sim=rank(maxsim))
# 
#   protal.rank.df <- join.gstd.rank %>% filter(Tool == "protal")
#   protal.rank.df <- protal.rank.df %>% arrange(sim)
#   protal.rank.df$sliding_window <- c(rep(NA, window_size/2), unlist(lapply(seq(nrow(protal.rank.df)-window_size), FUN=function(i) {
#     return(mean(protal.rank.df$Value[i:(i+window_size)]))
#   })), rep(NA, window_size/2))
# 
#   if ("StrainPhlAn4" %in% join.gstd.rank$Tool) {
#     strainphlan.rank.df <- join.gstd.rank %>% filter(Tool == "StrainPhlan4")
#     strainphlan.rank.df <- strainphlan.rank.df %>% arrange(sim)
#     strainphlan.rank.df$sliding_window <- c(rep(NA, window_size/2), unlist(lapply(seq(nrow(strainphlan.rank.df)-window_size), FUN=function(i) {
#       return(mean(strainphlan.rank.df$Value[i:(i+window_size)]))
#     })), rep(NA, window_size/2))
#     rank.df <- rbind(protal.rank.df, strainphlan.rank.df)
#   } else {
#     rank.df <- protal.rank.df
#   }
# 
#   window.plot <- rank.df %>%
#     ggplot(aes(x=sim, y=sliding_window, color=Tool)) +
#     geom_line() +
#     scale_x_continuous(labels=function(x){
#       return(round(sort(rank.df$maxsim)[x], digits=8))
#       }, breaks=as.integer(seq(0,1, 0.1)*(nrow(rank.df)-1))+1) +
#     theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
#     xlab("Closest neighboring genome") +
#     ylab("Monophyly per genome")
# 
#   return(window.plot)
# }
# 
# swp <- sliding_window_mono(join.gstd, 100)
# sliding_window_mono(join.gstd %>% filter(Tool != "gold_std" & Value != 1), 100)
# 
# 
# join %>% filter(Tool != "gold_std") %>%
#   ggplot(aes(x=maxsim, y=Value, color=Tool)) + 
#   geom_point(alpha=0.4)

```

## Tree completeness
```{r, echo=FALSE, warnings=FALSE, message=FALSE, eval=FALSE}
# Measure how many tips are in each of the trees? If protal e.g. can recover more samples in a tree then it has the edge over metaphlan.

map$tip.num <- unlist(lapply(map$Name, FUN=function(x) {length(tree.dict[[x]]$tip.label)}))

# Map species to total sample count
gold.std <- map %>% filter(Tool == "gold_std") %>% group_by(Species) %>% head()
species_to_tips <- gold.std$tip.num
names(species_to_tips) <- gold.std$Species
map <- map %>% mutate(tip.sensitivity=tip.num/species_to_tips[Species])

map %>% ggplot(aes(x=Tool, y=tip.sensitivity)) +
  geom_boxplot()


```

## Correlation and Tree distance analysis { .tabset }

```{r, echo=FALSE, results="asis", warnings=FALSE, eval=FALSE}
if (length(tree_dict_shared_tips) > 0) {
  mantel.m <- get_pairwise_mantel(data$map, data$tree.dict.shared.tips)
  mantel.m <- mantel.m %>% 
    left_join(map, by="Name")
  
  mantel.summary <- mantel.m %>% group_by(Tool, TreeTool) %>%
    summarise(PearsonCorMean=mean(Cor), PearsonCorSd=sd(Cor, na.rm = TRUE)) %>%
    as.data.frame()
  
  mantel.plot <- mantel.m %>% 
    filter(Species %in% data$shared_species) %>%
    ggplot(aes(x=Tool, y=Cor, fill=Tool)) + 
    geom_boxplot() +
    xlab("Tool") +
    ylab("Correlation")
  print(mantel.plot)
  
  # Summary plot of correlations
  correlation.summary <- mantel.m %>% filter(TreeTool == "iqtree") %>%
    ggplot(aes(x=Tool, y=Cor, fill=Tool)) +
    geom_boxplot() + ggtitle("Matrix Correlation")

  
  tree.pairwise <- get_pairwise_tree_distances(data$map, data$tree.dict.shared.tips) %>% 
    left_join(map, by="Name")
  
  # treedist.summary.plot <- tree.pairwise %>% filter(TreeTool == "iqtree") %>% 
  #   pivot_longer(cols=head(tail(names(tree.pairwise), -2), 4), names_to = "metric") %>%
  #   ggplot(aes(x=Tool, y=value, fill=Tool)) +
  #   geom_boxplot() +
  #   facet_wrap(~ metric, scales="free_y") + ggtitle("Tree Distances")
  # print(treedist.summary.plot)
  # 
  # tree_dist.summary <- tree.pairwise %>% group_by(Tool, TreeTool) %>%
  #   summarise(RobinsonFouldMean=mean(symmetric.difference), RobinsonFouldSd=sd(symmetric.difference, na.rm = TRUE),
  #             QuadraticPathMean=mean(quadratic.path.difference), QuadraticPathSd=sd(quadratic.path.difference, na.rm = TRUE)) %>%
  #   as.data.frame()
  
} else {
  
}


```


## {-}

```{r}
# 
# for (name in map$Name) {
#   print(name)
#   tree_meta <- meta.dict[[subdf[subdf$Name == name, MAP_META_COL]]]
#   tree.plot <- ggtree(tree.dict[[name]], layout='circular') %<+% tree_meta  +
#     geom_tippoint(aes(color=.data[[META_GROUP_COL]]), size=3) +
#     geom_tiplab(linesize=3, size=3, offset=0.001)+ theme(legend.position="None")
#   print(tree.plot)
# }
```

